### PR TITLE
couchbase: wait until all services are part of the config

### LINKED ISSUE #2993: CouchbaseContainer does not actually wait until the query service is available
It seems that even though the `CouchbaseContainer` checks the health of the `/admin/ping` endpoint for the query service, the query service is not actually available to handle queries yet. 

If I `withReuse(true)` this container, and run the test a second time, it succeeds because it now seems the query service has had enough time to initialize and become available to handle queries.

```
2020-07-14 13:26:20.936  INFO 80842 --- [           main] 🐳 [couchbase:enterprise-6.5.1]          : Creating container for image: couchbase:enterprise-6.5.1
2020-07-14 13:26:21.051  INFO 80842 --- [           main] 🐳 [couchbase:enterprise-6.5.1]          : Starting container with ID: 1ece82baa13e1965812f1371320d97b75e92fb2aab727246dd5e3ccba984a3d4
2020-07-14 13:26:21.629  INFO 80842 --- [           main] 🐳 [couchbase:enterprise-6.5.1]          : Container couchbase:enterprise-6.5.1 is starting: 1ece82baa13e1965812f1371320d97b75e92fb2aab727246dd5e3ccba984a3d4
2020-07-14 13:26:21.640  INFO 80842 --- [           main] o.t.c.wait.strategy.HttpWaitStrategy     : /youthful_dirac: Waiting for 60 seconds for URL: http://localhost:33429/pools
2020-07-14 13:26:21.661  INFO 80842 --- [tream-142261320] docker.couchbase                         : STDOUT: Starting Couchbase Server -- Web UI available at http://<ip>:8091
2020-07-14 13:26:21.661  INFO 80842 --- [tream-142261320] docker.couchbase                         : STDOUT: and logs available in /opt/couchbase/var/lib/couchbase/logs
2020-07-14 13:26:31.307  INFO 80842 --- [     ducttape-0] o.t.c.wait.strategy.HttpWaitStrategy     : /youthful_dirac: Waiting for 30 seconds for URL: http://localhost:33429/pools/default
2020-07-14 13:26:31.959  INFO 80842 --- [     ducttape-0] o.t.c.wait.strategy.HttpWaitStrategy     : /youthful_dirac: Waiting for 30 seconds for URL: http://localhost:33427/admin/ping
2020-07-14 13:26:32.970  INFO 80842 --- [           main] 🐳 [couchbase:enterprise-6.5.1]          : Container couchbase:enterprise-6.5.1 started in PT14.344S
2020-07-14 13:26:32.977  INFO 80842 --- [           main] o.t.c.wait.strategy.HttpWaitStrategy     : /youthful_dirac: Waiting for 60 seconds for URL: http://localhost:33429/pools/default/buckets/YS_Apps
2020-07-14 13:26:34.124  INFO 80842 --- [           main] o.t.c.wait.strategy.HttpWaitStrategy     : /youthful_dirac: Waiting for 60 seconds for URL: http://localhost:33429/pools/default/buckets/YS_Secure
2020-07-14 13:26:35.105  INFO 80842 --- [           main] o.t.c.wait.strategy.HttpWaitStrategy     : /youthful_dirac: Waiting for 60 seconds for URL: http://localhost:33429/pools/default/buckets/YS_Ephemeral
2020-07-14 13:26:36.145  INFO 80842 --- [           main] 🐳 [couchbase:enterprise-6.5.1]          : Couchbase container is ready! UI available at http://localhost:33429
```

```
2020-07-14 13:26:38.041  INFO 80842 --- [           main] com.couchbase.client.core.CouchbaseCore  : CouchbaseEnvironment: {sslEnabled=false, sslKeystoreFile='null', sslTruststoreFile='null', sslKeystorePassword=false, sslTruststorePassword=false, sslKeystore=null, sslTruststore=null, sslHostnameVerificationEnabled=false, bootstrapHttpEnabled=true, bootstrapCarrierEnabled=true, bootstrapHttpDirectPort=33429, bootstrapHttpSslPort=18091, bootstrapCarrierDirectPort=33422, bootstrapCarrierSslPort=11207, ioPoolSize=12, computationPoolSize=12, responseBufferSize=16384, requestBufferSize=16384, kvServiceEndpoints=1, viewServiceEndpoints=12, queryServiceEndpoints=12, searchServiceEndpoints=12, configPollInterval=2500, configPollFloorInterval=50, networkResolution=NetworkResolution{name='auto'}, ioPool=NioEventLoopGroup, kvIoPool=null, viewIoPool=null, searchIoPool=null, queryIoPool=null, analyticsIoPool=null, coreScheduler=CoreScheduler, memcachedHashingStrategy=DefaultMemcachedHashingStrategy, eventBus=DefaultEventBus, packageNameAndVersion=couchbase-java-client/2.7.11 (git: 2.7.11, core: 1.7.11), retryStrategy=BestEffort, maxRequestLifetime=75000, retryDelay=ExponentialDelay{growBy 1.0 MICROSECONDS, powers of 2; lower=100, upper=100000}, reconnectDelay=ExponentialDelay{growBy 1.0 MILLISECONDS, powers of 2; lower=32, upper=4096}, observeIntervalDelay=ExponentialDelay{growBy 1.0 MICROSECONDS, powers of 2; lower=10, upper=100000}, keepAliveInterval=30000, continuousKeepAliveEnabled=true, keepAliveErrorThreshold=4, keepAliveTimeout=2500, autoreleaseAfter=2000, bufferPoolingEnabled=true, tcpNodelayEnabled=true, mutationTokensEnabled=false, socketConnectTimeout=1000, callbacksOnIoPool=false, disconnectTimeout=25000, requestBufferWaitStrategy=com.couchbase.client.core.env.DefaultCoreEnvironment$4@4f654cee, certAuthEnabled=false, coreSendHook=null, forceSaslPlain=false, compressionMinRatio=0.83, compressionMinSize=32, compressionEnabled=true, operationTracingEnabled=true, operationTracingServerDurationEnabled=true, tracer=ThresholdLogTracer, orphanResponseReportingEnabled=true, orphanResponseReporter=DefaultOrphanResponseReporter, keyValueServiceConfig=KeyValueServiceConfig{minEndpoints=1, maxEndpoints=1, pipelined=true, idleTime=0}, queryServiceConfig=QueryServiceConfig{minEndpoints=0, maxEndpoints=12, pipelined=false, idleTime=300}, searchServiceConfig=SearchServiceConfig{minEndpoints=0, maxEndpoints=12, pipelined=false, idleTime=300}, viewServiceConfig=ViewServiceConfig{minEndpoints=0, maxEndpoints=12, pipelined=false, idleTime=300}, analyticsServiceConfig=AnalyticsServiceConfig{minEndpoints=0, maxEndpoints=12, pipelined=false, idleTime=300}, queryTimeout=7500, viewTimeout=7500, searchTimeout=75000, analyticsTimeout=75000, kvTimeout=2500, connectTimeout=5000, dnsSrvEnabled=false, propagateParentSpan=true}
2020-07-14 13:26:38.686  INFO 80842 --- [      cb-io-1-1] com.couchbase.client.core.node.Node      : Connected to Node localhost
2020-07-14 13:26:38.737  INFO 80842 --- [-computations-8] c.c.c.core.config.ConfigurationProvider  : Selected network configuration: external
2020-07-14 13:26:38.746  INFO 80842 --- [-computations-8] c.c.c.core.config.ConfigurationProvider  : Opened bucket YS_Apps
2020-07-14 13:26:38.752  INFO 80842 --- [      cb-io-1-1] com.couchbase.client.core.node.Node      : Disconnected from Node localhost
2020-07-14 13:26:38.771  INFO 80842 --- [      cb-io-1-2] com.couchbase.client.core.node.Node      : Connected to Node 172.18.0.2 (localhost)
```

```
org.springframework.dao.DataAccessResourceFailureException: The Query service is not enabled or no node in the cluster supports it.; nested exception is com.couchbase.client.core.ServiceNotAvailableException: The Query service is not enabled or no node in the cluster supports it.

	at org.springframework.data.couchbase.core.CouchbaseExceptionTranslator.translateExceptionIfPossible(CouchbaseExceptionTranslator.java:87)
	at org.springframework.data.couchbase.core.CouchbaseTemplate.execute(CouchbaseTemplate.java:551)
	at org.springframework.data.couchbase.core.CouchbaseTemplate.queryN1QL(CouchbaseTemplate.java:505)
```

### PR DESCRIPTION
**Pull request by daschl**
This changeset adds a predicate to the wait strategy to make sure
that it not only returns a 200, but also that every enabled service
is actually already exposed in the config. Since we are polling
the server during bootstrap here, not all of them might show up
at the same time.

Also, while not contributing to the fix we poll the terse bucket
http config "b" instead of the verbose one "buckets" since it is
a little more efficient on the server side and actually the config
the client internally works with.

fixes #2993

### CODE REVIEW

File: modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java [hunk 1]
```diff
@@ -450,4 +449,41 @@ private Response doHttpRequest(final int port, final String path, final String m
...
+                        }
+                        if (!found) {
+                            logger().trace("Service {} not yet part of config, retrying.", enabledService);
+                            return false;
+                        }
+                    }
+                }
+                return true;
+            } catch (IOException ex) {
+                logger().error("Unable to parse response {}", rawConfig);
```
Comment (aaronjwhiteside) [inline modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java:484]: exception is swallowed here

File: modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java [hunk 2]
```diff
@@ -450,4 +449,41 @@ private Response doHttpRequest(final int port, final String path, final String m
...
+     * <p>
+     *  Fixes the issue observed in https://github.com/testcontainers/testcontainers-java/issues/2993
+     */
+    private class AllServicesEnabledPredicate implements Predicate<String> {
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public boolean test(final String rawConfig) {
+            try {
+                Map<String, Object> parsedConfig = MAPPER.readValue(rawConfig, new TypeReference<Map<String, Object>>() {});
```
Comment (bsideup) [inline modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java:465]: Consider using `readTree`

Comment (aaronjwhiteside) [reply modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java:465]: I second this, less casting to Map/List..

Comment (daschl) [reply modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java:465]: I'll get it patched up shortly

File: modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java [hunk 3]
```diff
@@ -168,7 +174,7 @@ protected void configure() {
                             .map("healthy"::equals)
                             .orElse(false);
                     } catch (IOException e) {
-                        logger().error("Unable to parse response {}", response);
+                        logger().error("Unable to parse response: " + response, e);
```
Comment (aaronjwhiteside) [inline modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java:177]: put back `{}` for response, `Throwables` can be the last argument to `error()` calls.

File: modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java [hunk 4]
```diff
@@ -450,4 +452,38 @@ private Response doHttpRequest(final int port, final String path, final String m
...
+                        }
+                        if (!found) {
+                            logger().trace("Service {} not yet part of config, retrying.", enabledService);
+                            return false;
+                        }
+                    }
+                }
+                return true;
+            } catch (IOException ex) {
+                logger().error("Unable to parse response: " + rawConfig, ex);
```
Comment (aaronjwhiteside) [inline modules/couchbase/src/main/java/org/testcontainers/couchbase/CouchbaseContainer.java:484]: use `{}` placeholder for `rawConfig`

**Comment by daschl** [comment]
@bsideup @rnorth I think this should be good now, let me know if further modifications are needed :)

**Comment by aaronjwhiteside** [comment]
@daschl can please have the conflict resolved?

**Comment by aaronjwhiteside** [comment]
@daschl bump!

**Comment by daschl** [comment]
@aaronjwhiteside conflicts resolved and rebased with master.

**Comment by aaronjwhiteside** [comment]
thanks @daschl, much appreciated! 

@bsideup @rnorth does this PR require any further changes?

**Comment by aaronjwhiteside** [comment]
@bsideup @rnorth Is there any way I can help get this PR reviewed and merged?

**Comment by bsideup** [comment]
@aaronjwhiteside on it 👍

**Pull request merged**